### PR TITLE
Api create bookclub program

### PR DESCRIPTION
**Pull request by goelsonali**
Swagger contract - Programme API 
Screenshot 2024-08-06 at 00 06 14

API response 

Screenshot 2024-08-06 at 00 09 23

### CODE REVIEW

File: config/pmd/custom-ruleset.xml [hunk 1]
```diff
@@ -48,4 +48,12 @@
     </properties>
   </rule>
   <!--  END CODE DOCUMENTATION -->
+
+  <!-- DESIGN -->
+  <rule ref="rulesets/design.xml/ExcessiveParameterList">
+    <properties>
+      <property name="maxParameters" value="10"/>
```
Comment (womencodingcommunity) [inline config/pmd/custom-ruleset.xml:55]: I think would be better to add ignore for this check in the line this is happening now, instead of just allow 10 params, what do you think? Because for constructor we need that, but only there. And we can disable pmd only in the contructor line.

File: src/main/java/com/wcc/platform/domain/platform/EventSection.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package com.wcc.platform.domain.platform;
+
+import com.wcc.platform.domain.cms.attributes.SimpleLink;
+import java.util.List;
+import lombok.Builder;
+import lombok.Data;
+
```
Comment (dricazenck) [inline src/main/java/com/wcc/platform/domain/platform/EventSection.java:7]: missing javadoc

File: src/main/java/com/wcc/platform/domain/platform/Programme.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package com.wcc.platform.domain.platform;
+
+import com.wcc.platform.domain.cms.attributes.Card;
+import lombok.Builder;
+import lombok.Data;
+
+@Data
```
Comment (dricazenck) [inline src/main/java/com/wcc/platform/domain/platform/Programme.java:7]: missing javadoc

File: src/main/java/com/wcc/platform/service/CmsService.java [hunk 4]
```diff
@@ -16,86 +16,84 @@
...
-            return objectMapper.readValue(file, TeamPage.class);
-        } catch (IOException e) {
-            throw new PlatformInternalException(e.getMessage(), e);
-        }
+  /**
+   * Read JSON and convert to Pojo TeamPage.
+   *
+   * @return Pojo TeamPage.
+   */
+  public TeamPage getTeam() {
```
Comment (dricazenck) [inline src/main/java/com/wcc/platform/service/CmsService.java:35]: please revert this changes, because they have to be as string for now because of docker limitiation.. unless you manage to fix that. I mentioned in the deployment PR that getFileUri does not work on docker deployment

File: src/main/java/com/wcc/platform/service/ProgrammeService.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+  /**
+   * API to fetch details about different type of programmes.
+   *
+   * @return Programme Page json response
+   */
+  public ProgrammePage getProgramme(final ProgramType programType) {
+    try {
+      String data = null;
+      if (ProgramType.BOOK_CLUB.equals(programType)) {
```
Comment (dricazenck) [inline src/main/java/com/wcc/platform/service/ProgrammeService.java:33]: you don't need the programType as parameter here, we need validation in the controller level (which can be added later) 

but in the service we know that is for bookClub

File: src/main/resources/eventsPage.json [hunk 6]
```diff
@@ -28,13 +28,11 @@
       "startDate": "THU, MAY 30, 2024, 8:00 PM CEST",
       "endDate": "THU, MAY 30, 2024, 9.30 CEST",
       "title": "Book Club: The Pragmatic Programmer",
-      "speaker": "Jane Doe",
-      "speakerLink": {
+      "speakerProfile": {
```
Comment (dricazenck) [inline src/main/resources/eventsPage.json:31]: please fix the conflicts you have after I merged the fix in events.

File: src/test/java/com/wcc/platform/controller/ProgrammeControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.test.web.servlet.MockMvc;
+
+/** Unit test for the programme apis. */
+@WebMvcTest(ProgrammeController.class)
+public class ProgrammeControllerTest {
+  public static final String API_PROGRAMME = "/api/cms/v1/programme";
+  public static final String PROG_TYPE_BOOK_CLUB = "?type=BOOK_CLUB";
```
Comment (dricazenck) [inline src/test/java/com/wcc/platform/controller/ProgrammeControllerTest.java:27]: this can be private.

File: src/test/java/com/wcc/platform/factories/SetupEventFactories.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package com.wcc.platform.factories;
+
+import com.wcc.platform.domain.cms.attributes.ProgramType;
+import com.wcc.platform.domain.cms.attributes.SimpleLink;
+import com.wcc.platform.domain.platform.Event;
+
+/** Event test factories. */
+public class SetupEventFactories {
```
Comment (dricazenck) [inline src/test/java/com/wcc/platform/factories/SetupEventFactories.java:8]: sorry, I also create factories, basically you can remove this or copy my changes to yours factory and delete the other one file

File: src/test/java/com/wcc/platform/service/ProgrammeServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.wcc.platform.domain.cms.attributes.ProgramType;
+import com.wcc.platform.domain.cms.pages.programme.ProgrammePage;
+import com.wcc.platform.domain.exceptions.PlatformInternalException;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+
+/** Test class for ProgrammeService. */
+public class ProgrammeServiceTest {
```
Comment (dricazenck) [inline src/test/java/com/wcc/platform/service/ProgrammeServiceTest.java:20]:
```suggestion
class ProgrammeServiceTest {
```

**Review by dricazenck — APPROVED** [review]
Just small details, but once fixed you can merge :)

File: src/main/java/com/wcc/platform/service/ProgrammeService.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        data = FileUtil.readFileAsString(PROG_BOOK_CLUB.getFileName());
+      }
+      return objectMapper.readValue(data, ProgrammePage.class);
+    } catch (JsonProcessingException e) {
+      throw new PlatformInternalException(e.getMessage(), e);
```
Comment (dricazenck) [inline src/main/java/com/wcc/platform/service/ProgrammeService.java:38]: also please remember to add the missing test for the events. I tried to add some, but I was not able to finish all.

File: src/main/java/com/wcc/platform/controller/ProgrammeController.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+  @Autowired
+  public ProgrammeController(final ProgrammeService programmeService) {
+    this.service = programmeService;
+  }
+
+  @GetMapping("/api/cms/v1/programme")
+  @Operation(summary = "API to retrieve programme page")
+  @ResponseStatus(HttpStatus.OK)
+  public ResponseEntity<ProgrammePage> getProgramme(
+      @RequestParam(required = false, name = "type") final ProgramType programmeType) {
```
Comment (dricazenck) [inline src/main/java/com/wcc/platform/controller/ProgrammeController.java:32]:
```suggestion
      @RequestParam(required = true, name = "type") final ProgramType programmeType) {
```

File: src/main/java/com/wcc/platform/service/ProgrammeService.java (hunk 5 above)
Comment (goelsonali) [reply src/main/java/com/wcc/platform/service/ProgrammeService.java:33]: @dricazenck - I was reading through the best practices, I suppose the programme type validation should happen at the service layer rather than at the controller level - https://www.baeldung.com/spring-service-layer-validation
What do you think ?

File: src/main/java/com/wcc/platform/controller/ProgrammeController.java (hunk 11 above)
Comment (goelsonali) [reply src/main/java/com/wcc/platform/controller/ProgrammeController.java]: Done

File: config/pmd/custom-ruleset.xml (hunk 1 above)
Comment (goelsonali) [reply config/pmd/custom-ruleset.xml]: Removed this change to the rule and added the ignore statement to the cnstructor

File: src/main/java/com/wcc/platform/service/ProgrammeService.java (hunk 5 above)
Comment (goelsonali) [reply src/main/java/com/wcc/platform/service/ProgrammeService.java:33]: We can discuss and decide if this needs to be changed.
